api/policy.go

```diff
@@ -97,19 +100,22 @@ This would prevent someone from toggling something to disabled or 0 respectively
 As such we are using pointers instead of primitives for booleans and integers in this struct
 */
 type UpdatePolicy struct {
```

Comment:
We should validate what types are optional here as some of them may need to be pointer types.  For instance, if someone has set `AlertProfile`, how do they properly update a policy to "empty out" the `AlertProfile` (since `omitempty` will simply omit `AlertProfile`)...?

Reply:
Interesting. We should not infer empty strings and we should not omit empty strings in any of the policy apis (IMHO). So the fields in the struct probably needs to look like:
```golang
PolicyID *string `json:"policyId,omitempty"`
```
to be able to parse the user provided text and faithfully pass it on to the api (the existing `string` is broken as it will remove the `policyId` key from the json object if `PolicyID` is the empty string, and thus the customer may get an odd error message that is pertains to input that is different from what he wrote). In other words I would expect the struct to be:
```golang
type NewPolicy struct {
        PolicyID              *string   `json:"policyId,omitempty" yaml:"policyId,omitempty" `
        PolicyType            *string   `json:"policyType,omitempty" yaml:"policyType,omitempty"`
        QueryID               *string   `json:"queryId,omitempty" yaml:"queryId,omitempty"`
        Title                 *string   `json:"title,omitempty" yaml:"title,omitempty"`
        Enabled               *bool     `json:"enabled,omitempty" yaml:"enabled,omitempty"`
        Description           *string   `json:"description,omitempty" yaml:"description,omitempty"`
        Remediation           *string   `json:"remediation,omitempty" yaml:"remediation,omitempty"`
        Severity              *string   `json:"severity,omitempty" yaml:"severity,omitempty"`
        Limit                 *int      `json:"limit,omitempty" yaml:"limit,omitempty"`
        EvalFrequency         *string   `json:"evalFrequency,omitempty" yaml:"evalFrequency,omitempty"`
        AlertEnabled          *bool     `json:"alertEnabled,omitempty" yaml:"alertEnabled,omitempty"`
        AlertProfile          *string   `json:"alertProfile,omitempty" yaml:"alertProfile,omitempty"`
        Tags                  []string `json:"tags,omitempty" yaml:"tags,omitempty"`
        References            []string `json:"references,omitempty" yaml:"references,omitempty"`
        AdditionalInformation *string   `json:"additionalInformation,omitempty" yaml:"additionalInformation,omitempty"`
        InfoLink              *string   `json:"infoLink,omitempty" yaml:"infoLink,omitempty"`
}
```
to not mess with the input in a material way.

Such changes are outside the scope of what I want to fix here, so I'll just make sure to not expand the problem with the type of the new fields.

Reply:
The use of pointers is strictly related to the patching (`UpdatePolicy` struct) of optional values such that we can differentiate between zero vs. unspecified...